ecommerce/models.py

```diff
@@ -0,0 +1,271 @@
+    )
+    object_id = models.PositiveIntegerField()
+    content_object = GenericForeignKey("content_type", "object_id")
+
+    class Meta:
+        unique_together = ("content_type", "object_id")
+
+    def __str__(self):
+        """Description of a product"""
+        return f"Product for {self.content_object}"
+
+
+class ProductVersion(TimestampedModel):
+    """
+    An append-only table for Product, keeping track of the parts of a product which are editable.
+    """
+
+    product = models.ForeignKey(
+        Product, on_delete=models.PROTECT, related_name="productversions"
+    )
+    price = models.DecimalField(decimal_places=2, max_digits=20)
+    description = models.TextField()
+
+    class Meta:
+        indexes = [models.Index(fields=["created_on"])]
+
+    def __str__(self):
+        """Description of a ProductVersion"""
+        return f"ProductVersion for {self.description}, ${self.price}"
+
+
+class Basket(TimestampedModel):
+    """
+    Represents a User's basket
+    """
+
+    user = models.OneToOneField(settings.AUTH_USER_MODEL, on_delete=models.PROTECT)
+
+    def __str__(self):
+        """Description of Basket"""
+        return f"Basket for {self.user}"
+
+
+class BasketItem(TimestampedModel):
+    """
+    Represents a product in a user's basket
+    """
+
+    product = models.ForeignKey(Product, on_delete=models.PROTECT)
+    basket = models.ForeignKey(
+        Basket, on_delete=models.PROTECT, related_name="basketitems"
+    )
+    quantity = models.PositiveIntegerField()
+
+    def __str__(self):
+        """Description of BasketItem"""
+        return f"BasketItem of product {self.product} (qty: {self.quantity}"
+
+
+class Order(TimestampedModel, AuditableModel):
+    """
+    An order containing information for a purchase.
+    """
+
+    FULFILLED = "fulfilled"
+    FAILED = "failed"
+    CREATED = "created"
+    REFUNDED = "refunded"
+
+    STATUSES = [CREATED, FULFILLED, FAILED, REFUNDED]
+    FULFILLED_STATUSES = [FULFILLED]
+
+    purchaser = models.ForeignKey(
+        settings.AUTH_USER_MODEL, on_delete=models.PROTECT, related_name="orders"
+    )
+    status = models.CharField(
+        choices=[(status, status) for status in STATUSES],
+        default=CREATED,
+        max_length=30,
+        db_index=True,
+    )
+
+    def __str__(self):
+        """Description for Order"""
+        return f"Order #{self.id}, status={self.status}"
+
+    @classmethod
+    def get_audit_class(cls):
+        return OrderAudit
+
+    def to_dict(self):
+        """
+        Get a serialized representation of the Order and any attached Basket and Lines
+        """
+        raise NotImplementedError
+
+
+class OrderAudit(AuditModel):
+    """
+    Audit model for Order
+    """
+
+    order = models.ForeignKey(Order, null=True, on_delete=models.PROTECT)
+
+    @classmethod
+    def get_related_field_name(cls):
+        return "order"
```

Comment:
Would be nice to start the pattern of referring to the actual field property here

Reply:
I actually looked at this, I'm not sure how to do this for this particular field. Mind if I delay until next PR?